vcs/gitcmd/repo.go

```diff
@@ -161,105 +161,122 @@ func (r *Repository) ResolveTag(name string) (vcs.CommitID, error) {
 	return commitID, nil
 }
 
-// branchCounts returns the behind/ahead commit counts information for branch, against base branch.
-func (r *Repository) branchCounts(branch, base string) (behind, ahead uint, err error) {
-	if err := checkSpecArgSafety(branch); err != nil {
-		return 0, 0, err
-	}
-	if err := checkSpecArgSafety(base); err != nil {
-		return 0, 0, err
-	}
+// branchFilter is a filter for branch names.
+// If not empty, only contained branch names are allowed. If empty, all name are allowed.
+// The map should be made so it's not nil.
+type branchFilter map[string]struct{}
+
+// allows will return true if the current filter set-up validates against
+// the passed string. If there are no filters, all strings pass.
+func (f branchFilter) allows(name string) bool {
+	if len(f) == 0 {
+		return true
+	}
+	_, ok := f[name]
+	return ok
+}
 
-	cmd := exec.Command("git", "rev-list", "--count", "--left-right", fmt.Sprintf("refs/heads/%s...refs/heads/%s", base, branch))
-	cmd.Dir = r.Dir
-	out, err := cmd.Output()
-	if err != nil {
-		return 0, 0, err
+// add adds a slice of strings to the filter.
+func (f branchFilter) add(list []string) {
+	for _, l := range list {
+		f[l] = struct{}{}
 	}
-	behindAhead := strings.Split(strings.TrimSuffix(string(out), "\n"), "\t")
-	b, err := strconv.ParseUint(behindAhead[0], 10, 0)
-	if err != nil {
-		return 0, 0, err
-	}
-	a, err := strconv.ParseUint(behindAhead[1], 10, 0)
-	if err != nil {
-		return 0, 0, err
-	}
-	return uint(b), uint(a), nil
 }
 
 func (r *Repository) Branches(opt vcs.BranchesOptions) ([]*vcs.Branch, error) {
 	r.editLock.RLock()
 	defer r.editLock.RUnlock()
 
-	refs, err := r.showRef("--heads")
-	if err != nil {
-		return nil, err
-	}
-
-	branches := make([]*vcs.Branch, len(refs))
-	for i, ref := range refs {
-		branches[i] = &vcs.Branch{
-			Name: strings.TrimPrefix(ref[1], "refs/heads/"),
-			Head: vcs.CommitID(ref[0]),
+	f := make(branchFilter)
+	if opt.MergedInto != "" {
+		b, err := r.branches("--merged", opt.MergedInto)
```

Comment:
This could be left for another PR, but it seems a little odd that we're doing `"--merged", "<commit>"` and `"--contains=" + "<commit>"` - two different styles of arguments for `git branch` command.

According to http://git-scm.com/docs/git-branch, it looks like only `--merged <commit>` and `--contains <commit>` are supported. But both forms seem to work. Still, it'd be better to change to one style for consistency (I'm in favor of changing `--contains=` to be without `=` liked `--merged`.